{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Availability \n",
    "\n",
    "In this notebook, the proteins and genes extracted from the Proteomics.ipynb and RNA-seq.ipynb are vetted for aviaibaility. \n",
    "\n",
    "This was added due to inconsistent results in the Final notebook and was proven to be essential to the process. Genes that are not available for at least 72 out of 80 samples (90%) are discarded in the final classification step. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final Results.ipynb MSI Visual.ipynb    \u001b[34m__pycache__\u001b[m\u001b[m         \u001b[34mraw_data\u001b[m\u001b[m\r\n",
      "Gender Visual.ipynb Proteomics.ipynb    availability.ipynb  subchallenge_2.csv\r\n",
      "Import.ipynb        RNA-Seq.ipynb       preprocess.py\r\n"
     ]
    }
   ],
   "source": [
    "!ls\n",
    "#######################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "from IPython.utils import io\n",
    "from io import StringIO\n",
    "\n",
    "import re\n",
    "import csv\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy \n",
    "from scipy.cluster import hierarchy as hc # for dendograms \n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "%matplotlib inline\n",
    "# %matplotlib notebook\n",
    "\n",
    "from sklearn.metrics import f1_score # f1_score(y_true, y_pred)\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.model_selection import GridSearchCV\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.metrics import classification_report\n",
    "from sklearn.feature_selection import SelectFromModel"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import & show data \n",
    "\n",
    "The initial data proprocess was done in the Import.ipynb notebook. The following three steps are performed in all other jupyter notebooks: run the preprocess.py code that was generated in Import.ipynb, run the %whos magic command to see what dataframes have been imported, and then format the captured %who results with pandas to view results. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%run preprocess.py\n",
    "# made in Import.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "with io.capture_output() as captured:\n",
    "    %whos DataFrame \n",
    "\n",
    "s = re.findall(r'\\S+|\\n', StringIO(captured.stdout).getvalue())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variable</th>\n",
       "      <th>Type</th>\n",
       "      <th>Data/Info</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>df_test_cli</td>\n",
       "      <td>DataFrame</td>\n",
       "      <td>sample , gender , &lt;...&gt;g_80 , Female , MSI-High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>df_test_pro</td>\n",
       "      <td>DataFrame</td>\n",
       "      <td>A1BG , &lt;...&gt;n[80 , rows , x , 4118 , columns]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>df_test_rna</td>\n",
       "      <td>DataFrame</td>\n",
       "      <td>A1BG , A1B&lt;...&gt;[80 , rows , x , 17447 , columns]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>df_train_cli</td>\n",
       "      <td>DataFrame</td>\n",
       "      <td>sample , gender , &lt;...&gt;g_80 , Female , MSI-Low/MSS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>df_train_mislabel</td>\n",
       "      <td>DataFrame</td>\n",
       "      <td>sample , mismatch&lt;...&gt;79 , Training_80 , 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>df_train_pro</td>\n",
       "      <td>DataFrame</td>\n",
       "      <td>A1BG , &lt;...&gt;n[80 , rows , x , 4118 , columns]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>df_train_rna</td>\n",
       "      <td>DataFrame</td>\n",
       "      <td>A1BG , A1&lt;...&gt;[80 , rows , x , 17447 , columns]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>test_cli</td>\n",
       "      <td>DataFrame</td>\n",
       "      <td>gender , msi\\n&lt;...&gt;\\nTesting_80 , 1 , 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>test_pro</td>\n",
       "      <td>DataFrame</td>\n",
       "      <td>A1BG , &lt;...&gt;n[80 , rows , x , 4118 , columns]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>test_rna</td>\n",
       "      <td>DataFrame</td>\n",
       "      <td>A1BG , A1B&lt;...&gt;[80 , rows , x , 17447 , columns]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>train_cli</td>\n",
       "      <td>DataFrame</td>\n",
       "      <td>gender , msi\\n0 , &lt;...&gt; , 0 , 0\\n79 , 1 , 0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>train_combined</td>\n",
       "      <td>DataFrame</td>\n",
       "      <td>mismatch , gender , msi&lt;...&gt;[80 , rows , x , 21568 , columns]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>train_mislabel</td>\n",
       "      <td>DataFrame</td>\n",
       "      <td>mismatch\\n0 , &lt;...&gt;8 , 1\\n79 , 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>train_pro</td>\n",
       "      <td>DataFrame</td>\n",
       "      <td>A1BG , A2M , &lt;...&gt;n[80 , rows , x , 4118 , columns]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>train_pro_combined</td>\n",
       "      <td>DataFrame</td>\n",
       "      <td>mismatch , gender , msi&lt;...&gt;n[80 , rows , x , 4121 , columns]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>train_rna</td>\n",
       "      <td>DataFrame</td>\n",
       "      <td>A1BG , A1BG-AS1 , &lt;...&gt;[80 , rows , x , 17447 , columns]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>train_rna_combined</td>\n",
       "      <td>DataFrame</td>\n",
       "      <td>mismatch , gender , msi&lt;...&gt;[80 , rows , x , 17450 , columns]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Variable       Type  \\\n",
       "0   df_test_cli         DataFrame   \n",
       "1   df_test_pro         DataFrame   \n",
       "2   df_test_rna         DataFrame   \n",
       "3   df_train_cli        DataFrame   \n",
       "4   df_train_mislabel   DataFrame   \n",
       "5   df_train_pro        DataFrame   \n",
       "6   df_train_rna        DataFrame   \n",
       "7   test_cli            DataFrame   \n",
       "8   test_pro            DataFrame   \n",
       "9   test_rna            DataFrame   \n",
       "10  train_cli           DataFrame   \n",
       "11  train_combined      DataFrame   \n",
       "12  train_mislabel      DataFrame   \n",
       "13  train_pro           DataFrame   \n",
       "14  train_pro_combined  DataFrame   \n",
       "15  train_rna           DataFrame   \n",
       "16  train_rna_combined  DataFrame   \n",
       "\n",
       "                                                        Data/Info  \n",
       "0   sample , gender , <...>g_80 , Female , MSI-High                \n",
       "1   A1BG , <...>n[80 , rows , x , 4118 , columns]                  \n",
       "2   A1BG , A1B<...>[80 , rows , x , 17447 , columns]               \n",
       "3   sample , gender , <...>g_80 , Female , MSI-Low/MSS             \n",
       "4   sample , mismatch<...>79 , Training_80 , 1                     \n",
       "5   A1BG , <...>n[80 , rows , x , 4118 , columns]                  \n",
       "6   A1BG , A1<...>[80 , rows , x , 17447 , columns]                \n",
       "7   gender , msi\\n<...>\\nTesting_80 , 1 , 1                        \n",
       "8   A1BG , <...>n[80 , rows , x , 4118 , columns]                  \n",
       "9   A1BG , A1B<...>[80 , rows , x , 17447 , columns]               \n",
       "10  gender , msi\\n0 , <...> , 0 , 0\\n79 , 1 , 0                    \n",
       "11  mismatch , gender , msi<...>[80 , rows , x , 21568 , columns]  \n",
       "12  mismatch\\n0 , <...>8 , 1\\n79 , 1                               \n",
       "13  A1BG , A2M , <...>n[80 , rows , x , 4118 , columns]            \n",
       "14  mismatch , gender , msi<...>n[80 , rows , x , 4121 , columns]  \n",
       "15  A1BG , A1BG-AS1 , <...>[80 , rows , x , 17447 , columns]       \n",
       "16  mismatch , gender , msi<...>[80 , rows , x , 17450 , columns]  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table = []\n",
    "row = []\n",
    "\n",
    "for item in s:\n",
    "    if (item == '\\n'):\n",
    "        if (row):\n",
    "            table.append(row)\n",
    "        row = []\n",
    "    elif re.match('^(?!-)', item) is None:\n",
    "        pass        \n",
    "    else: \n",
    "        row.append(item)\n",
    "\n",
    "for item in table: \n",
    "    if (item[0] == 'Variable'):\n",
    "        pass\n",
    "    else:\n",
    "        item[2:] = [' , '.join(item[2:])]\n",
    "\n",
    "df = pd.DataFrame(data=table[1:], columns=table[0])\n",
    "pd.set_option('display.max_colwidth', -1)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# train_pro_combined\n",
    "# train_rna_combined\n",
    "# train_combined.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### These are extracted from Proteomics and RNA-Seq notebooks, many of the non-available genes were discarded in the porcess and these are only left now; meaning the current output does not reflect the process and work that was done to dicard those genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "TAP1        0\n",
       "TAP1        0\n",
       "APOL2       0\n",
       "APOL2       0\n",
       "LCP1        0\n",
       "LCP1        0\n",
       "PTPN6       0\n",
       "PTPN6       0\n",
       "CASK        0\n",
       "CASK        0\n",
       "UBE2L6      0\n",
       "UBE2L6      0\n",
       "ICAM1       0\n",
       "ICAM1       0\n",
       "ITGB2       0\n",
       "ITGB2       0\n",
       "HP          0\n",
       "HP          0\n",
       "HK3         0\n",
       "HK3         0\n",
       "RFC2        0\n",
       "RFC2        0\n",
       "WARS        0\n",
       "WARS        0\n",
       "SDF2L1      0\n",
       "SDF2L1      0\n",
       "CKB         0\n",
       "CKB         0\n",
       "LAP3        0\n",
       "LAP3        0\n",
       "IFI30       0\n",
       "IFI30       0\n",
       "PTPRC       0\n",
       "PTPRC       0\n",
       "HSDL2       0\n",
       "HSDL2       0\n",
       "IFI35       0\n",
       "IFI35       0\n",
       "TYMP        0\n",
       "TYMP        0\n",
       "CSRP2       0\n",
       "CSRP2       0\n",
       "TAPBP       0\n",
       "TAPBP       0\n",
       "ERMP1       0\n",
       "ERMP1       0\n",
       "ANP32E      0\n",
       "ANP32E      0\n",
       "ROCK2       0\n",
       "ROCK2       0\n",
       "           ..\n",
       "SRSF6       0\n",
       "GYG2P1      0\n",
       "DDB2        0\n",
       "ZRSR2       0\n",
       "EMX1        0\n",
       "EHBP1       0\n",
       "THG1L       0\n",
       "KCNJ2       0\n",
       "ZNF606      0\n",
       "EVPL        0\n",
       "LUC7L2      0\n",
       "LUC7L2      0\n",
       "SRGAP2B     0\n",
       "NAT8        0\n",
       "ZNF862      0\n",
       "ODAM        0\n",
       "RNASEH2A    0\n",
       "RNASEH2A    0\n",
       "EIF1AY      0\n",
       "EIF1AY      0\n",
       "FOXD2       0\n",
       "PMS2P3      0\n",
       "MND1        0\n",
       "FAM3D       0\n",
       "FAM3D       0\n",
       "DDX3X       0\n",
       "DDX3X       0\n",
       "NINJ2       0\n",
       "NLRX1       0\n",
       "NLRX1       0\n",
       "COCH        0\n",
       "IRF5        0\n",
       "FHDC1       0\n",
       "FBXW9       0\n",
       "RPL21P28    0\n",
       "TLX1        0\n",
       "HEBP2       0\n",
       "HEBP2       0\n",
       "TRIM29      0\n",
       "ZNF273      0\n",
       "CRABP2      0\n",
       "ABCA6       0\n",
       "MMP17       0\n",
       "PCNA        0\n",
       "PCNA        0\n",
       "S100A16     0\n",
       "S100A16     0\n",
       "UBA6        0\n",
       "UBA6        0\n",
       "RGS4        0\n",
       "Length: 242, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_combined[\n",
    "    \n",
    "    [ # msi proteomics\n",
    "    'TAP1', 'APOL2', 'LCP1', 'PTPN6', 'CASK', 'UBE2L6', \n",
    "    'ICAM1', 'ITGB2', 'HP', 'HK3', 'RFC2', 'WARS',\n",
    "    'SDF2L1', 'CKB', 'LAP3', 'IFI30', 'PTPRC', 'HSDL2', \n",
    "    'IFI35', 'TYMP', 'CSRP2', 'TAPBP', 'ERMP1', 'ANP32E', \n",
    "    'ROCK2', 'ADPGK', 'HSPA4L', 'CNDP2', 'RFTN1', 'GBP1', \n",
    "    'GBP2', 'FMNL1', 'DMBT1', 'ARL3',\n",
    "    'NCF2', 'YARS2', 'RPL3', 'SPTLC2', 'ENO1', 'SNX12', \n",
    "                \n",
    "    # mis rna-seq \n",
    "    'EPDR1', 'APOL3', 'POU5F1B', 'CFTR', 'CIITA', 'RAB32', \n",
    "    'MAX', 'TRIM72', 'GBP4', 'HPSE', 'WARS', 'GRM8', 'SLC51B',\n",
    "    'PRSS23', 'FABP6', 'GABRP', 'LAP3', 'LY6G6D', 'SLC19A3', \n",
    "    'DLGAP1-AS5', 'GBP1', 'RAMP1', 'AREG', 'GSPT2', 'TNFAIP2',\n",
    "    'EREG', 'TNNC2', 'ANKRD27', 'PLCL2', 'TFCP2L1', 'LAG3', \n",
    "    'BEX2', 'DEFB1', 'FABP1', 'IRF1', 'CCL4', 'SLC25A48', \n",
    "\n",
    "    # gender proteomics\n",
    "    'MYO9B', 'EMILIN1', 'MRPL16', 'EIF4G2', 'TNXB', 'ARFIP1', \n",
    "    'GAR1', 'SLC35A2', 'FBLN1', 'FLNC', 'S100A14', 'COX7C', \n",
    "    'FBN1', 'C4A','COL4A2','RCC2','ZNF706', 'NAP1L1', 'AGRN', \n",
    "    'GADD45GIP1', 'TJP2', 'ALDH1A1', 'PGM2', 'LTBP1', 'NNMT', \n",
    "    'SERPINB1', 'U2AF1L5', 'COG3', 'ASPN', 'METTL1', 'DDT', \n",
    "\n",
    "    # gender rna-seq \n",
    "    'RPS4Y1', 'ZFX', 'DDX3Y', 'UTY', 'SRSF6', 'GYG2P1', 'DDB2', \n",
    "    'ZRSR2', 'EMX1', 'EHBP1', 'THG1L', 'KCNJ2', 'ZNF606',\n",
    "    'EVPL', 'LUC7L2', 'SRGAP2B', 'NAT8', 'ZNF862', 'ODAM', \n",
    "    'RNASEH2A', 'EIF1AY', 'FOXD2', 'PMS2P3', 'MND1', 'FAM3D', \n",
    "    'DDX3X', 'NINJ2', 'NLRX1', 'COCH', 'IRF5', 'FHDC1', \n",
    "    'FBXW9', 'RPL21P28', 'TLX1', 'HEBP2', 'TRIM29', 'ZNF273', \n",
    "    'CRABP2', 'ABCA6', 'MMP17', 'PCNA', 'S100A16', 'UBA6', 'RGS4', \n",
    "    ]\n",
    "    \n",
    "            ].isnull().sum(axis = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
